### PR TITLE
SUBMARINE-202. submarine core need to support MXNet

### PR DESCRIPTION
**Pull request by lowc1012**
### What is this PR for?
To support MXNet framework in Submarine

### What type of PR is it?
[Improvement]

### Todos
* [ ] - Task

### What is the Jira issue?
[SUBMARINE-202](https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-202)

### How should this be tested?
[passed CI](https://travis-ci.org/lowc1012/submarine/builds/646785076)

### Screenshots (if appropriate)
screenshot1
screenshot2

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by pingsutw** [comment]
@lowc1012 Thanks for the contribution.
- We also need to add `RunJobCliParsingMxnetTest.java`, `RunJobCliParsingMxnetYamlTest.java` like `submarine/submarine-client/src/test/java/org/apache/submarine/client/cli/runjob/pytorch/`
https://github.com/apache/submarine/tree/master/submarine-client/src/test/java/org/apache/submarine/client/cli/runjob/pytorch
- This PR looks like we only support MXnet run on TonY runtime, we need to make sure it can also support RPC runtime and restful runtime, you could file another JIRA ticket

**Comment by yuanzac** [comment]
@lowc1012,
Thanks for the contributions. Can you take a look at the travis error?

**Comment by liuxunorg** [comment]
@lowc1012 Can you rebase the development branch and commit it?

Need to execute the following type command.
```
git rebase upstream/master
git push -f origin SUBMARINE-202:SUBMARINE-202
```

**Comment by liuxunorg** [comment]
@pingsutw @yuanzac @tangzhankun  Can you help review this PR?

File: submarine-client/src/main/java/org/apache/submarine/client/cli/runjob/RunJobCli.java [hunk 1]
```diff
@@ -219,6 +224,23 @@ private void addTensorboardOptions(Options options) {
...
+    options.addOption(CliConstants.SCHEDULER_DOCKER_IMAGE, true,
+        "Specify docker image for scheduler, when this is not specified, " +
+        "scheduler uses --" + CliConstants.DOCKER_IMAGE +
+        " as default. " + CAN_BE_USED_WITH_MXNET_ONLY);
+    options.addOption(CliConstants.SCHEDULER_LAUNCH_CMD, true,
+        "Commandline of scheduler, arguments will be " +
+        "directly used to launch the scheduler. " + CAN_BE_USED_WITH_MXNET_ONLY);
+    options.addOption(CliConstants.SCHEDULER_RES, true,
+        "Resource of each scheduler, for example " +
+        "memory-mb=2048,vcores=2,yarn.io/gpu=2. " + CAN_BE_USED_WITH_MXNET_ONLY);
```
Comment (pingsutw) [inline submarine-client/src/main/java/org/apache/submarine/client/cli/runjob/RunJobCli.java]: Will MXnet scheduler server use GPU? if not, we can remove `yarn.io/gpu=2`

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java [hunk 2]
```diff
@@ -54,67 +55,22 @@ public static Configuration tonyConfFromClientContext(
...
-      tonyConf.setLong(
-          TonyConfigurationKeys.getResourceKey(Constants.PS_JOB_NAME,
-              Constants.MEMORY),
-          ResourceUtils.getMemorySize(psResource));
-    }
-    if (parameters.getOptionValue(CliConstants.WORKER_RES) != null) {
-      Resource workerResource = getResource(parameters, CliConstants.WORKER_RES);
+    setParametersForWorker(tonyConf, parameters);
+    setParametersForPS(tonyConf, parameters);
+    setParametersForScheduler(tonyConf, parameters);
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: We could add a function to check if submarine job uses MXNET.
If we use MXNET and then we `setParametersForScheduler` for it.

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java [hunk 3]
```diff
@@ -203,4 +147,118 @@ private static Resource getResource(Parameter parametersHolder, String option)
...
+      tonyConf.setInt(
+              TonyConfigurationKeys.getResourceKey(Constants.SCHEDULER_JOB_NAME,
+                      Constants.VCORES),
+              schedulerResource.getVirtualCores());
+      tonyConf.setLong(
+              TonyConfigurationKeys.getResourceKey(Constants.SCHEDULER_JOB_NAME,
+                      Constants.MEMORY),
+              ResourceUtils.getMemorySize(schedulerResource));
+      tonyConf.setLong(
+              TonyConfigurationKeys.getResourceKey(Constants.SCHEDULER_JOB_NAME,
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: if MXNET scheduler don't use GPU, we should remove these lines as well.

File: submarine-client/src/main/java/org/apache/submarine/client/cli/runjob/RunJobCli.java (hunk 1 above)
Comment (lowc1012) [reply submarine-client/src/main/java/org/apache/submarine/client/cli/runjob/RunJobCli.java]: Thanks for the review! I will remove it.

File: submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java (hunk 3 above)
Comment (lowc1012) [reply submarine-server/server-submitter/submitter-yarn/src/main/java/org/apache/submarine/server/submitter/yarn/YarnUtils.java]: Thanks for the review! I will remove it.

**Comment by yuanzac** [comment]
LGTM~
@lowc1012,
Thanks for the contributions. Have you tested the MXNet job on yarn cluster? 
With this PR, we can run MXNet job with both command parameters and yaml parameter, Right?

**Comment by yuanzac** [comment]
Any more comments? @lowc1012, @pingsutw, @liuxunorg, @tangzhankun

**Comment by lowc1012** [comment]
@yuanzac, @pingsutw 
Thanks for your review. I have tested on my YARN cluster.
Yes, we can submit MXNet job with this PR.
And I made some modifications, please review it . Thank you!

**Comment by yuanzac** [comment]
@lowc1012，
Thanks, LGTM.
@pingsutw, The modification of YarnUtils may conflicts with your PR #179. Can you help to take a look.
Thanks

**Comment by pingsutw** [comment]
+1, LGTM. Thanks @lowc1012 
@yuanzac He just adds a `setParametersForScheduler(tonyConf, parameters);` for MXNET
After it merges, I will rebase PR #179 and fix conflict

**Pull request closed without merging**